libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -202,93 +242,79 @@ bool AC_WPNav::get_wp_destination(Location& destination) const
     return true;
 }
 
-/// set_wp_destination waypoint using position vector (distance from home in cm)
-///     terrain_alt should be true if destination.z is a desired altitude above terrain
+/// set_wp_destination - set destination waypoints using position vectors (distance from ekf origin in cm)
+///     terrain_alt should be true if origin.z and destination.z are desired altitudes above terrain (false if these are alt-above-ekf-origin)
+///     returns false on failure (likely caused by missing terrain data)
 bool AC_WPNav::set_wp_destination(const Vector3f& destination, bool terrain_alt)
 {
-	Vector3f origin;
+    if (is_active() && _flags.reached_destination && (terrain_alt == _terrain_alt)) {
+        // use previous destination as origin
+        _origin = _destination;
 
-    // if waypoint controller is active use the existing position target as the origin
-    if ((AP_HAL::millis() - _wp_last_update) < 1000) {
-        origin = _pos_control.get_pos_target();
+        // store previous leg
+        _scurve_prev_leg = _scurve_this_leg;
     } else {
-        // if waypoint controller is not active, set origin to reasonable stopping point (using curr pos and velocity)
-        _pos_control.get_stopping_point_xy(origin);
-        _pos_control.get_stopping_point_z(origin);
-    }
 
-    // convert origin to alt-above-terrain
-    if (terrain_alt) {
-        float origin_terr_offset;
-        if (!get_terrain_offset(origin_terr_offset)) {
-            return false;
+        // use stopping point as the origin
+        get_wp_stopping_point(_origin);
+
+        // convert origin to alt-above-terrain if necessary
+        if (terrain_alt) {
+            float origin_terr_offset;
+            if (!get_terrain_offset(origin_terr_offset)) {
+                return false;
```

Comment:
AT the bottom of set_wp_destination() we zap the next_leg by setting these values which ensures the next_leg will not be used and if it were, it would not contribute back any desired position, velocity or acceleration changes.

- _this_leg_is_spline = false;
- _scurve_next_leg.init();
- _flags.fast_waypoint = false;